arangod/RestHandler/RestIndexHandler.cpp

```diff
@@ -199,7 +205,102 @@ RestStatus RestIndexHandler::getIndexes() {
+              futures.reserve(shards->size());
+              std::string const prefix = "/_api/index/";
+              network::RequestOptions reqOpts;
+              reqOpts.param("withHidden", "true");
+              reqOpts.database = _vocbase.name();
+              // best effort. only displaying progress
+              reqOpts.timeout = network::Timeout(10.0);
+              for (auto const& shard : *shards) {
+                std::string const url =
+                    absl::StrCat(prefix, shard.first, "/", iid);
+                futures.emplace_back(network::sendRequestRetry(
+                    pool, "shard:" + shard.first, fuerte::RestVerb::Get, url,
+                    body, reqOpts));
+              }
+              for (Future<network::Response>& f : futures) {
+                network::Response const& r = f.get();
+
+                // Only best effort accounting. If something breaks here, we
+                // just ignore the output. Account for what we can and move on.
+                if (r.fail()) {
+                  LOG_TOPIC("afde4", INFO, Logger::CLUSTER)
+                      << "Communication error while fetching index data "
+                         "for collection "
+                      << coll->name() << " from " << r.destination;
+                  continue;
+                }
+                VPackSlice resSlice = r.slice();
+                if (!resSlice.isObject() ||
+                    !resSlice.get(StaticStrings::Error).isBoolean()) {
+                  LOG_TOPIC("aabe4", INFO, Logger::CLUSTER)
+                      << "Result of collecting index data for collection "
+                      << coll->name() << " from " << r.destination
+                      << " is invalid";
+                  continue;
+                }
+                if (resSlice.get(StaticStrings::Error).getBoolean()) {
+                  LOG_TOPIC("a4bea", INFO, Logger::CLUSTER)
+                      << "Failed to collect index data for collection "
+                      << coll->name() << " from " << r.destination;
+                  continue;
+                }
+                if (resSlice.get("progress").isNumber()) {
+                  progress += resSlice.get("progress").getNumber<double>();
+                } else {
+                  // Obviously, the index is already ready there.
+                  progress += 100.0;
+                  LOG_TOPIC("aeab4", DEBUG, Logger::CLUSTER)
+                      << "No progress entry on index " << iid << "  from "
+                      << r.destination << ": " << resSlice.toJson()
+                      << " index already finished.";
+                }
+              }
+              if (progress != 0) {
+                // Don't show progress 0, this is in particular relevant
+                // when isBackground is false, in which case no progress
+                // is reported by design.
+                tmp.add("progress", VPackValue(progress / shards->size()));
```

Comment:
nit: if one of the shards fails to answer, dividing by `shards->size()` may underreport the actual progress.
I don't have a better solution here, except counting the number of shards for which we successfully received a response. But as the progress value may be skewed anyway in this case, it does not matter much.

Reply:
This is correct, but unlikely to occur. Dividing only by the number of correct results also has issues. Let's leave it like it is for now.